pkg/traceql/combine.go

```diff
@@ -149,29 +149,42 @@ type QueryRangeCombiner struct {
 	req     *tempopb.QueryRangeRequest
 	eval    *MetricsFrontendEvaluator
 	metrics *tempopb.SearchMetrics
+
+	maxSeries        int
+	maxSeriesReached bool
+
+	// used to track which series were updated since the previous diff
+	// todo: it may not be worth it to track the diffs per series. it would be simpler (and possibly nearly as effective) to just calculate a global
+	//  max/min for all series
+	seriesUpdated map[string]tsRange
 }
 
-func QueryRangeCombinerFor(req *tempopb.QueryRangeRequest, mode AggregateMode) (*QueryRangeCombiner, error) {
+func QueryRangeCombinerFor(req *tempopb.QueryRangeRequest, mode AggregateMode, maxSeries int) (*QueryRangeCombiner, error) {
 	eval, err := NewEngine().CompileMetricsQueryRangeNonRaw(req, mode)
 	if err != nil {
 		return nil, err
 	}
 
 	return &QueryRangeCombiner{
-		req:     req,
-		eval:    eval,
-		metrics: &tempopb.SearchMetrics{},
+		req:       req,
+		eval:      eval,
+		maxSeries: maxSeries,
+		metrics:   &tempopb.SearchMetrics{},
 	}, nil
 }
 
 func (q *QueryRangeCombiner) Combine(resp *tempopb.QueryRangeResponse) {
-	if resp == nil {
+	if resp == nil || q.maxSeriesReached {
 		return
 	}
 
 	// Here is where the job results are reentered into the pipeline
 	q.eval.ObserveSeries(resp.Series)
 
+	if q.maxSeries > 0 && len(q.eval.Results()) >= q.maxSeries {
```

Comment:
i'm guessing "Results()" can get quite expensive. it might be easier to count and limit input series? this would be similar in concept to limiting input streams in a loki or prometheus query. can you do some research here to determine the cost?

Reply:
switching to return series count in `ObserveSeries`